compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -4089,6 +4089,7 @@ public void visit(BLangBinaryExpr binaryExpr) {
         SymbolEnv rhsExprEnv;
         BType lhsType;
         if (binaryExpr.expectedType.tag == TypeTags.FLOAT || binaryExpr.expectedType.tag == TypeTags.DECIMAL ||
+                                                             binaryExpr.expectedType.tag == TypeTags.BYTE ||
```

Comment:
I don't think this is the correct fix for the original issue.

Check https://github.com/ballerina-platform/ballerina-spec/issues/997#issuecomment-976383649. 

```ballerina
byte x = 10 & 10;
```

should work because the operands are singletons and therefore the result being a singleton (related lang issue: https://github.com/ballerina-platform/ballerina-lang/issues/33090). And that singleton being a subtype of the expected type. 

I think fixing #33090 should fix this issue.